proposals/0000-transient-storages.md

```diff
@@ -0,0 +1,336 @@
+"Right now we abuse storage for intra-block data such as block number, parent hash and block author as well as various housekeeping information and flags like whether we set the uncles Authorship::DidSetUncles.
+
+When initially writing, this incurs an extra trie lookup, which is slow. Instead there should be another host API, which works exactly like set_storage/get_storage but has no trie backing, so it never tries to lookup the value in the trie, nor does it write the value at the end of the block." Credits Gavin Wood.
+
+Additionally content such as events often are used in a log based manner (append only) with possibly a larger size than usual content.
+
+Hashing through host function involves passing all data at once, and is not memory efficient.
+
+## Implementation
+
+Transient storage act as current state storage, but without a persistent backend.
+
+This implies that the storage must support commiting and reverting transaction with `ext_storage_commit_transaction` or `ext_storage_rollback_transaction`.
+This transactional support is both at transient storage content and at transient storage definition (a delete transient storage will be restore on rollback).
+
+Btree and blob are using a specific `Mode`, either `drop` or `archive` passed respectively as the byte 0 or 1. When using `drop` the data will not be send from the runtime executor to the calling client. When using `archive` the committed state of the transient storage will be passed as a change set to the client calling runtime executor.
+
+In archive mode it is the client that choose its strategy for storing the block transient storages final states.
+
+
+### Implementation of Btree storage
+
+- `ext_btree_storage_new` with parameters:
+	- name : a pointer size to the name of the new transient storage. 
+	- mode : `Mode` as an u8 (either 0 `drop` or 1 `archive).
+No result.
+Allows using a transient storage for a given `name` and `mode`.
+If a transient storage already exists with the same `name`, it is overwritten.
+
+- `ext_btree_storage_exists` with parameters:
+	- name : a pointer size to the name of a transient storage.
+Result is a boolean indicating if transient storage was instantiated.
+
+- `ext_btree_storage_delete` with parameters:
+	- name : a pointer size to the name of a transient storage.
+Result true if a transient storage did exist and was removed, and false if no
+transient storage did exist.
+
+
+- `ext_btree_storage_clone` with parameters:
+	- name : a pointer size to the name of a transient storage to clone.
+	- target_name : a pointer size to the new transient storage to use.
+Result is a true if operation succeed and false if there was no storage at `name`.
+Clone keep same `Mode`. Clone copy all content from a storage to another storage.
+If a transient storage is present at `target_name` it is overwritten.
+
+This operation cost is high, the implementation do not try to avoid copy.
+
+- `ext_btree_storage_rename` with parameters:
```

Comment:
Why is this function necessary? Especially if transient storages aren't stored across blocks, I have trouble seeing why you would need to rename a storage.

Reply:
I initially thought only to have a file like api, but thought of a possible usage.

You got a blob define globally: let s say system pallet defines "log" blob.
Then every extrinsic can access it and append to it.
But if a pallet want to use code from other pallet but not appending to log, then it just need to rename "log" do something on another "log" and afterward restore "log", at pretty much 0 cost.
So in this case to isolate a storage and restore it afterward. Same thing with potentially switching context for the remaining of a block processing.

But I would not say it is strictly necessary. And considering it is the more involved implementation detail (to keep it 0 cost), I would not drop the idea of not having it.